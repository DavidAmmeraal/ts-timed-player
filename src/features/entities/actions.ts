/**
 * actions.ts
 * Action creators for entities actions.
 */
import Types from 'Types';
import { createAction } from 'typesafe-actions';

export const createEntityAction = createAction(
  'entities/CREATE_ENTITY',
  resolve => (entityType: string, props: Types.EntityProps) =>
    resolve({
      entityType,
      props,
    }),
);

export const updateEntityAction = createAction(
  'entities/UPDATE_ENTITY',
  resolve => (entityType: string, props: Types.EntityProps) =>
    resolve({
      entityType,
      props,
    }),
);

export const deleteEntityAction = createAction(
  'entities/DELETE_ENTITY',
  resolve => (entityType: string, id: string) =>
    resolve({
      entityType,
      id,
    }),
);
