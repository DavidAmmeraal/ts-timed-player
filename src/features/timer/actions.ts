/**
 * actions.ts
 * Action creators for timer.
 */
import { action } from 'typesafe-actions';

import { TOGGLE, TICK, SET_RATE, SET_TIME } from './constants';

export const toggle = () => action(TOGGLE);
export const tick = (time: number) => action(TICK, time);
export const setRate = (rate: number) => action(SET_RATE, rate);
/**
 * Creates an SET_TIME time action.
 * @param time time in microseconds
 */
export const setTime = (time: number) => action(SET_TIME, time);
