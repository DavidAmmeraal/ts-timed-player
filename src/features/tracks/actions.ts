import { action } from 'typesafe-actions';

import { CREATE_TRACK, UPDATE_TRACK, DELETE_TRACK } from './constants';
import { ITrack } from './models';

export const createTrack = (track:ITrackEntity) => action(CREATE_TRACK, track);
export const updateTrack = (id:string, track: Partial<ITrack>) => action(UPDATE_TRACK, {
    id,
    track,
});
export const deleteTrack = (id:string) => action(DELETE_TRACK, id);
