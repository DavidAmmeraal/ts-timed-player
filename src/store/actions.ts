import { createAsyncAction } from 'typesafe-actions';

import { FETCH_STAGE_REQUEST, FETCH_STAGE_SUCCESS, FETCH_STAGE_ERROR } from './constants';

export const fetchStage = createAsyncAction(
  FETCH_STAGE_REQUEST,
  FETCH_STAGE_SUCCESS,
  FETCH_STAGE_ERROR,
)<void, {}, Error>();
