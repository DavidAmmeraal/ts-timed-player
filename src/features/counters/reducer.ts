import { ActionType } from 'typesafe-actions';
import * as actions from './actions';
import { ADD, INCREMENT } from './constants';

export type CountersAction = ActionType<typeof actions>;
export type CountersState = {
  readonly value: number;
};

export const initialState: CountersState = {
  value: 1,
};

export default (state: CountersState = initialState, action: CountersAction) => {
  switch (action.type) {
    case ADD:
      return {
        ...state,
        value: state.value + action.payload,
      };
    case INCREMENT:
      return {
        ...state,
        value: state.value + 1,
      };
    default:
      return state;
  }
};
