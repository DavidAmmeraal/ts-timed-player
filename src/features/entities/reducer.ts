import * as actions from './actions';
import { ActionType } from 'typesafe-actions';
import { EntitiesContainer, createEntitiesContainer } from './models';
import { CREATE, UPDATE, DELETE } from './constants';
import { pipe, unless, has, lensProp, set, lensPath, over, union } from 'ramda';

export type EntitiesAction = ActionType<typeof actions>;
type CreateAction = ActionType<typeof actions.createEntityAction>;

export interface EntitiesState {
    [key: string]: EntitiesContainer,
}
export const defaultState: EntitiesState = {};

/**
 * Handles entities/CREATE actions. Creates EntitiesContainer for entity type if it doesn't yet. 
 * 
 * @param state Base state.
 * @param action entities/CREATE action to be handled.
 * @returns new state.
 */
const handleCreate = (state:EntitiesState, { payload: { entityType, entity }}:CreateAction) => pipe
    <EntitiesState, EntitiesState, EntitiesState, EntitiesState>(
    unless(
        has(entityType), 
        set(lensProp(entityType), createEntitiesContainer()),
    ),
    set(lensPath([entityType, 'byId', entity.id]), entity),
    over(lensPath([entityType, 'ids']), union([entity.id])),
)(state);

/**
 * Handles and reduces actions into state. 
 * 
 * @param state The current state.
 * @param action The action to be handled.
 */
const reducer = (state: EntitiesState = defaultState, action: EntitiesAction) => {
    switch (action.type) {
        case CREATE:
            return handleCreate(state, action);
        case UPDATE:
            return state;
        case DELETE:
            return state;
        default:
            return state;
    }
};
export default reducer;
