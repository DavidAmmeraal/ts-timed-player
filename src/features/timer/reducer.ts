import * as actions from './actions';
import { ActionType } from 'typesafe-actions';
import { TOGGLE, TICK, SET_RATE, SET_TIME } from './constants';

export type TimerAction = ActionType<typeof actions>;

export type TimerState = {
  readonly time: number;
  readonly rate: number;
  readonly running: boolean;
};

export const defaultState: TimerState = {
  time: 0,
  rate: 1,
  running: false,
};

export default (state: TimerState = defaultState, action: TimerAction) => {
  switch (action.type) {
    case TOGGLE:
      return {
        ...state,
        running: !state.running,
      };
    case TICK:
      return {
        ...state,
        time: state.time + state.rate * action.payload,
      };
    case SET_RATE:
      return {
        ...state,
        rate: action.payload,
      };
    case SET_TIME:
      return {
        ...state,
        time: action.payload,
      };
    default:
      return state;
  }
};
