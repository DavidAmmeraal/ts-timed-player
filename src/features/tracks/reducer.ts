import * as actions from './actions';

export default createEntitiesReducer({
    create: typeof actions.createTrack
})