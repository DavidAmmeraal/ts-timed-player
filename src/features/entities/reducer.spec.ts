import reducer from './reducer';
import * as actions from './actions';

/**
 * FIXTURES
 */
const initialState = reducer(undefined, {} as any);

describe('models reducer', () => {
    it('should have empty initial state', () => {
        expect(initialState).toMatchSnapshot();
    })
    describe('handling CREATE action', () => {
        it('should create a new entity of given type, and create a new type container if it does not exist.', () => {
            const entity = { id: '123', foo: 'bar' };
            const action = actions.createEntityAction('SuperCoolEntity', entity);
            const state = reducer(undefined, action);
            expect(state).toMatchSnapshot();
        })

        it('should create a new entity of given type and add it to the existing type container.', () => {
            const entity1 = { id: '123', foo: 'bar' };
            const action1 = actions.createEntityAction('SuperCoolEntity', entity1);
            const state = reducer(undefined, action1);

            const entity2 = { id: '432', bar: 'foo' };
            const action2 = actions.createEntityAction('SuperCoolEntity', entity2);

            const finalState = reducer(state, action2);
            expect(finalState).toMatchSnapshot();
        })

        it('should create a new entity of new type and add it next to other types.', () => {
            const superCoolEntity = { id: '123', foo: 'bar' };
            const superCoolAction = actions.createEntityAction('SuperCoolEntity', superCoolEntity);
            const state = reducer(undefined, superCoolAction);

            const superLameEntity = { id: '432', bar: 'foo' };
            const superLameAction = actions.createEntityAction('SuperLameEntity', superLameEntity);

            const finalState = reducer(state, superLameAction);
            expect(finalState).toMatchSnapshot();
        })

        it('should overwrite existing entities of same type if it has same id.', () => {
            const superCoolEntity = { id: '123', foo: 'bar', bar: 'foo' };
            const superCoolAction = actions.createEntityAction('SuperCoolEntity', superCoolEntity);
            const state = reducer(undefined, superCoolAction);

            const superCoolEntity2 = { id: '123', foo: 'foo' }
            const superCoolAction2 = actions.createEntityAction('SuperCoolEntity', superCoolEntity2);

            const finalState = reducer(state, superCoolAction2);
            expect(finalState).toMatchSnapshot();
        })
    })
    
});