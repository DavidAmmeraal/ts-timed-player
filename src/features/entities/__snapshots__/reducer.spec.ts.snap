// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models reducer handling CREATE action should create a new entity of given type and add it to the existing type container. 1`] = `
Object {
  "SuperCoolEntity": Object {
    "byId": Object {
      "123": Object {
        "foo": "bar",
        "id": "123",
      },
      "432": Object {
        "bar": "foo",
        "id": "432",
      },
    },
    "ids": Array [
      "432",
      "123",
    ],
  },
}
`;

exports[`models reducer handling CREATE action should create a new entity of given type, and create a new type container if it does not exist. 1`] = `
Object {
  "SuperCoolEntity": Object {
    "byId": Object {
      "123": Object {
        "foo": "bar",
        "id": "123",
      },
    },
    "ids": Array [
      "123",
    ],
  },
}
`;

exports[`models reducer handling CREATE action should create a new entity of new type and add it next to other types. 1`] = `
Object {
  "SuperCoolEntity": Object {
    "byId": Object {
      "123": Object {
        "foo": "bar",
        "id": "123",
      },
    },
    "ids": Array [
      "123",
    ],
  },
  "SuperLameEntity": Object {
    "byId": Object {
      "432": Object {
        "bar": "foo",
        "id": "432",
      },
    },
    "ids": Array [
      "432",
    ],
  },
}
`;

exports[`models reducer handling CREATE action should overwrite existing entities of same type if it has same id. 1`] = `
Object {
  "SuperCoolEntity": Object {
    "byId": Object {
      "123": Object {
        "foo": "foo",
        "id": "123",
      },
    },
    "ids": Array [
      "123",
    ],
  },
}
`;

exports[`models reducer should have empty initial state 1`] = `Object {}`;
